File: test/resolver_transport_contract.ml
type test = string * (unit -> unit)

exception Failed of string

let fail format = Printf.ksprintf (fun value -> raise (Failed value)) format
let expect message condition = if not condition then fail "%s" message

let dependency ?(locator = Frontend_intf.Direct_reference) provider kind
    reference =
  {
    Frontend_intf.provider;
    kind;
    reference;
    locator;
    span = Span.none;
    mutability = Frontend_intf.Mutable;
    status = Frontend_intf.Unresolved (Unknown.Unresolved_dependency reference);
  }

let response ?(status = 200) url body =
  {
    Resolver_transport.status;
    body;
    effective_url = url;
    peer_ip = "93.184.216.34";
  }

let github_action_test () =
  let revision = String.make 40 'a' and requests = ref [] in
  let get (request : Resolver_transport.request) =
    requests := request :: !requests;
    if
      Util.starts_with
        ~prefix:"https://api.github.com/repos/actions/checkout/commits/"
        request.url
    then Ok (response request.url (revision ^ "\n"))
    else if
      request.url
      = "https://codeload.github.com/actions/checkout/tar.gz/" ^ revision
    then Ok (response request.url "immutable archive")
    else if
      request.url
      = "https://raw.githubusercontent.com/actions/checkout/" ^ revision
        ^ "/action.yml"
    then
      Ok
        (response request.url
           "name: checkout\nruns:\n  using: node20\n  main: dist/index.js\n")
    else Error ("unexpected URL " ^ request.url)
  in
  let network = Resolver_transport.make ~get ~allowed_sources:[] in
  let fetched =
    match
      network.Resolver.fetch
        (dependency Ir.Github Frontend_intf.Action "actions/checkout@v4")
    with
    | Ok value -> value
    | Error message -> fail "%s" message
  in
  expect "GitHub ref resolves to an exact commit" (fetched.revision = revision);
  expect "the complete immutable source archive is hashed"
    (fetched.content = "immutable archive");
  expect "source provenance names the immutable tree"
    (fetched.source = "https://github.com/actions/checkout/tree/" ^ revision);
  expect "exact action metadata is retained for semantic inference"
    (match fetched.semantic_source with
    | Some source ->
        source.path = "action.yml"
        && Util.contains ~needle:"node20" source.content
    | None -> false);
  expect "resolution performs commit, archive, and metadata requests"
    (List.length !requests = 3)

let gitlab_component_test () =
  let revision = String.make 40 'b' and requests = ref [] in
  let get (request : Resolver_transport.request) =
    requests := request.url :: !requests;
    if Util.contains ~needle:"/repository/commits/1.2.3" request.url then
      Ok (response request.url ("{\"id\":\"" ^ revision ^ "\"}"))
    else if Util.contains ~needle:"/repository/archive.tar.gz?sha=" request.url
    then Ok (response request.url "gitlab archive")
    else if
      Util.contains ~needle:"templates%2Faws%2Ftemplate.yml/raw" request.url
    then Ok (response request.url "deploy:\n  script: echo exact\n")
    else Error ("unexpected URL " ^ request.url)
  in
  let network = Resolver_transport.make ~get ~allowed_sources:[] in
  let fetched =
    match
      network.fetch
        (dependency Ir.Gitlab Frontend_intf.Component
           "gitlab.com/acme/components/deploy/aws@1.2.3")
    with
    | Ok value -> value
    | Error message -> fail "%s" message
  in
  expect "GitLab component resolves its project commit"
    (fetched.revision = revision);
  expect "GitLab component snapshots the project archive"
    (fetched.content = "gitlab archive");
  expect "GitLab component retains its exact template source"
    (Option.is_some fetched.semantic_source);
  expect "GitLab resolver performs commit, archive, and template requests"
    (List.length !requests = 3)

let gitlab_project_file_test () =
  let revision = String.make 40 'd' and requests = ref [] in
  let get (request : Resolver_transport.request) =
    requests := request.url :: !requests;
    if Util.contains ~needle:("/repository/commits/" ^ revision) request.url
    then Ok (response request.url ("{\"id\":\"" ^ revision ^ "\"}"))
    else if Util.contains ~needle:("archive.tar.gz?sha=" ^ revision) request.url
    then Ok (response request.url "project include archive")
    else if
      Util.contains ~needle:"repository/files/templates%2Fbuild.yml/raw"
        request.url
    then Ok (response request.url "build:\n  script: echo exact\n")
    else Error ("unexpected URL " ^ request.url)
  in
  let network = Resolver_transport.make ~get ~allowed_sources:[] in
  let fetched =
    match
      network.fetch
        (dependency
           ~locator:
             (Frontend_intf.Repository_file
                {
                  repository = "acme/ci";
                  revision = Some revision;
                  path = "/templates/build.yml";
                  repository_type = None;
                })
           Ir.Gitlab Frontend_intf.Repository
           ("acme/ci:/templates/build.yml@" ^ revision))
    with
    | Ok value -> value
    | Error message -> fail "%s" message
  in
  expect "GitLab project include resolves the declared immutable revision"
    (fetched.revision = revision);
  expect "GitLab project include provenance identifies the exact file"
    (Util.ends_with ~suffix:(revision ^ "/templates/build.yml") fetched.source);
  expect "GitLab project file retains exact semantic source"
    (match fetched.semantic_source with
    | Some source -> source.path = "templates/build.yml"
    | None -> false);
  expect "GitLab project file adds one exact raw-file request"
    (List.length !requests = 3)

let azure_task_test () =
  let revision = String.make 40 'c' and requests = ref [] in
  let get (request : Resolver_transport.request) =
    requests := request.url :: !requests;
    if
      Util.contains ~needle:"/microsoft/azure-pipelines-tasks/commits/main"
        request.url
    then Ok (response request.url revision)
    else if Util.contains ~needle:("/tar.gz/" ^ revision) request.url then
      Ok (response request.url "azure task archive")
    else if
      Util.contains
        ~needle:(revision ^ "/Tasks/UsePythonVersionV0/task.json")
        request.url
    then
      Ok
        (response request.url
           "{\"execution\":{\"Node20_1\":{\"target\":\"main.js\"}}}")
    else Error ("unexpected URL " ^ request.url)
  in
  let network = Resolver_transport.make ~get ~allowed_sources:[] in
  let fetched =
    match
      network.fetch
        (dependency Ir.Azure Frontend_intf.Task "UsePythonVersion@0")
    with
    | Ok value -> value
    | Error message -> fail "%s" message
  in
  expect "Azure task source is commit pinned" (fetched.revision = revision);
  expect "Azure task provenance includes its task directory"
    (Util.ends_with
       ~suffix:(revision ^ "/Tasks/UsePythonVersionV0")
       fetched.source);
  expect "Azure task retains exact task.json metadata"
    (match fetched.semantic_source with
    | Some source ->
        Util.ends_with ~suffix:"task.json" source.path
        && Util.contains ~needle:"Node20_1" source.content
    | None -> false);
  expect "Azure task performs commit, archive, and metadata requests"
    (List.length !requests = 3)

let azure_github_template_test () =
  let revision = String.make 40 'e' and requests = ref [] in
  let get (request : Resolver_transport.request) =
    requests := request.url :: !requests;
    if
      Util.contains ~needle:"/repos/org/templates/commits/refs%2Ftags%2Fv1"
        request.url
    then Ok (response request.url revision)
    else if Util.ends_with ~suffix:("/tar.gz/" ^ revision) request.url then
      Ok (response request.url "GitHub template archive")
    else if Util.ends_with ~suffix:(revision ^ "/shared.yml") request.url then
      Ok (response request.url "steps:\n  - script: echo exact\n")
    else Error ("unexpected URL " ^ request.url)
  in
  let network = Resolver_transport.make ~get ~allowed_sources:[] in
  let fetched =
    match
      network.fetch
        (dependency
           ~locator:
             (Frontend_intf.Repository_file
                {
                  repository = "org/templates";
                  revision = Some "refs/tags/v1";
                  path = "shared.yml";
                  repository_type = Some "github";
                })
           Ir.Azure Frontend_intf.Template "shared.yml@shared")
    with
    | Ok value -> value
    | Error message -> fail "%s" message
  in
  expect "Azure GitHub template alias resolves through an exact commit"
    (fetched.revision = revision);
  expect "Azure template provenance includes the selected path"
    (Util.ends_with ~suffix:(revision ^ "/shared.yml") fetched.source);
  expect "Azure GitHub template retains exact template source"
    (match fetched.semantic_source with
    | Some source -> source.path = "shared.yml"
    | None -> false);
  expect "Azure GitHub template adds one exact raw-file request"
    (List.length !requests = 3)

let azure_native_template_test () =
  let revision = String.make 40 'f' and requests = ref [] in
  let get (request : Resolver_transport.request) =
    requests := request.url :: !requests;
    if Util.contains ~needle:"/commits?" request.url then
      Ok
        (response request.url
           ("{\"value\":[{\"commitId\":\"" ^ revision ^ "\"}]}"))
    else if Util.contains ~needle:"recursionLevel=Full" request.url then
      Ok (response request.url "Azure Repos archive")
    else if Util.contains ~needle:"path=%2Fshared.yml" request.url then
      Ok (response request.url "steps:\n  - script: echo exact\n")
    else Error ("unexpected URL " ^ request.url)
  in
  let network = Resolver_transport.make ~get ~allowed_sources:[] in
  let fetched =
    match
      network.fetch
        (dependency
           ~locator:
             (Frontend_intf.Repository_file
                {
                  repository = "org/project/templates";
                  revision = Some "refs/tags/v1";
                  path = "shared.yml";
                  repository_type = Some "azurereposgit";
                })
           Ir.Azure Frontend_intf.Template "shared.yml@templates")
    with
    | Ok value -> value
    | Error message -> fail "%s" message
  in
  expect "native Azure template resolves to an exact commit"
    (fetched.revision = revision);
  expect "native Azure template retains exact file content"
    (match fetched.semantic_source with
    | Some source ->
        source.path = "shared.yml"
        && Util.contains ~needle:"echo exact" source.content
    | None -> false);
  expect "native Azure template performs commit, archive, and file requests"
    (List.length !requests = 3)

let circleci_orb_test () =
  let version_id = "7a09fb7b-4415-4aee-bc0f-a2f7f8395824" in
  let get (request : Resolver_transport.request) =
    if Util.contains ~needle:"/api/v3/orb/versions?filter%5Bref%5D=" request.url
    then
      Ok
        (response request.url
           ("{\"data\":[{\"attributes\":{\"version\":\"5.0.3\"},\"id\":\""
          ^ version_id ^ "\"}]}"))
    else if Util.ends_with ~suffix:(version_id ^ "/source") request.url then
      Ok (response request.url "version: 2.1\ncommands: {}\n")
    else Error ("unexpected URL " ^ request.url)
  in
  let network = Resolver_transport.make ~get ~allowed_sources:[] in
  let fetched =
    match
      network.fetch
        (dependency Ir.Circleci Frontend_intf.Orb "circleci/node@5.0.3")
    with
    | Ok value -> value
    | Error message -> fail "%s" message
  in
  expect "CircleCI production orb retains its immutable SemVer"
    (fetched.revision = "5.0.3");
  expect "CircleCI source endpoint supplies the lock content"
    (Util.contains ~needle:"commands" fetched.content);
  expect "CircleCI orb source is also semantic evidence"
    (match fetched.semantic_source with
    | Some source -> source.path = ".circleci/config.yml"
    | None -> false)

let direct_and_network_boundary_test () =
  let calls = ref 0 in
  let get (request : Resolver_transport.request) =
    incr calls;
    Ok (response request.url "remote include")
  in
  let network =
    Resolver_transport.make ~get
      ~allowed_sources:[ "https://ci.example.test/includes/" ]
  in
  let fetched =
    match
      network.fetch
        (dependency Ir.Gitlab Frontend_intf.Include
           "https://ci.example.test/includes/base.yml")
    with
    | Ok value -> value
    | Error message -> fail "%s" message
  in
  expect "direct include is locked by its exact content digest"
    (Util.starts_with ~prefix:"sha256:" fetched.revision);
  expect "direct include content is available to semantic inference"
    (Option.is_some fetched.semantic_source);
  let blocked =
    network.fetch
      (dependency Ir.Gitlab Frontend_intf.Include "https://127.0.0.1/internal")
  in
  expect "untrusted network origins fail before transport"
    (Result.is_error blocked && !calls = 1);
  let redirecting request =
    Ok
      {
        Resolver_transport.status = 200;
        body = "stolen";
        effective_url = "https://evil.example/redirect";
        peer_ip = "93.184.216.34";
      }
  in
  let redirected =
    (Resolver_transport.make ~get:redirecting
       ~allowed_sources:[ "https://ci.example.test/" ])
      .fetch
      (dependency Ir.Gitlab Frontend_intf.Include
         "https://ci.example.test/base.yml")
  in
  expect "redirects are checked against the same source boundary"
    (Result.is_error redirected)

let tests =
  [
    ("GitHub actions resolve to immutable source trees", github_action_test);
    ( "GitLab components resolve to immutable source trees",
      gitlab_component_test );
    ( "GitLab project includes resolve typed repository files",
      gitlab_project_file_test );
    ("Azure tasks resolve to the official immutable task tree", azure_task_test);
    ( "Azure template aliases resolve typed GitHub files",
      azure_github_template_test );
    ( "Azure Repos templates retain exact semantic source",
      azure_native_template_test );
    ( "CircleCI production orbs resolve through the source API",
      circleci_orb_test );
    ( "resolver transport enforces the network origin boundary",
      direct_and_network_boundary_test );
  ]

let () =
  let failures = ref 0 in
  List.iter
    (fun (name, run) ->
      try
        run ();
        Printf.printf "ok - %s\n%!" name
      with
      | Failed message ->
          incr failures;
          Printf.eprintf "not ok - %s: %s\n%!" name message
      | error ->
          incr failures;
          Printf.eprintf "not ok - %s: unexpected %s\n%!" name
            (Printexc.to_string error))
    tests;
  if !failures > 0 then exit 1

File: crates/cli/tests/resolver_transport_contract.rs
use workflow_verifier_cli::resolver_transport::{ResolverRequest, resolve_dependency};
use workflow_verifier_domain::Provider;
use workflow_verifier_foundation::Span;
use workflow_verifier_frontend::{Dependency, DependencyKind, DependencyLocator};

fn dependency(provider: Provider, kind: DependencyKind, reference: &str) -> Dependency {
    Dependency::unresolved(
        provider,
        kind,
        reference,
        DependencyLocator::Direct,
        Span::default(),
    )
}

fn located_dependency(
    provider: Provider,
    kind: DependencyKind,
    reference: &str,
    locator: DependencyLocator,
) -> Dependency {
    Dependency::unresolved(provider, kind, reference, locator, Span::default())
}

#[test]
fn github_action_resolves_commit_archive_and_exact_metadata() {
    let revision = "a".repeat(40);
    let mut requests = Vec::new();
    let mut get = |request: &ResolverRequest| {
        requests.push(request.clone());
        if request.url.contains("/repos/actions/checkout/commits/v4") {
            Ok(format!("{revision}\n").into_bytes())
        } else if request
            .url
            .ends_with(&format!("/actions/checkout/tar.gz/{revision}"))
        {
            Ok(b"immutable archive".to_vec())
        } else if request
            .url
            .ends_with(&format!("/actions/checkout/{revision}/action.yml"))
        {
            Ok(b"name: checkout\nruns:\n  using: node20\n  main: dist/index.js\n".to_vec())
        } else {
            Err(format!("unexpected URL {}", request.url))
        }
    };

    let fetched = resolve_dependency(
        &dependency(
            Provider::Github,
            DependencyKind::Action,
            "actions/checkout@v4",
        ),
        &mut get,
    )
    .expect("resolve GitHub action");

    assert_eq!(fetched.revision, revision);
    assert_eq!(fetched.content, b"immutable archive");
    assert_eq!(
        fetched
            .semantic_source
            .as_ref()
            .map(|value| value.0.as_str()),
        Some("action.yml")
    );
    assert_eq!(requests.len(), 3);
    assert!(
        requests[0]
            .headers
            .iter()
            .any(|(name, value)| name == "Accept" && value == "application/vnd.github.sha")
    );
}

#[test]
fn gitlab_component_resolves_project_archive_and_template() {
    let revision = "b".repeat(40);
    let mut requests = Vec::new();
    let mut get = |request: &ResolverRequest| {
        requests.push(request.url.clone());
        if request.url.contains("/repository/commits/1.2.3") {
            Ok(format!("{{\"id\":\"{revision}\"}}").into_bytes())
        } else if request.url.contains("/repository/archive.tar.gz?sha=") {
            Ok(b"gitlab archive".to_vec())
        } else if request.url.contains("templates%2Faws%2Ftemplate.yml/raw") {
            Ok(b"deploy:\n  script: echo exact\n".to_vec())
        } else {
            Err(format!("unexpected URL {}", request.url))
        }
    };

    let fetched = resolve_dependency(
        &dependency(
            Provider::Gitlab,
            DependencyKind::Component,
            "gitlab.com/acme/components/deploy/aws@1.2.3",
        ),
        &mut get,
    )
    .expect("resolve GitLab component");

    assert_eq!(fetched.revision, revision);
    assert_eq!(fetched.content, b"gitlab archive");
    assert_eq!(
        fetched
            .semantic_source
            .as_ref()
            .map(|value| value.0.as_str()),
        Some("templates/aws/template.yml")
    );
    assert_eq!(requests.len(), 3);
}

#[test]
fn azure_task_resolves_official_task_tree_and_metadata() {
    let revision = "c".repeat(40);
    let mut requests = Vec::new();
    let mut get = |request: &ResolverRequest| {
        requests.push(request.url.clone());
        if request
            .url
            .contains("/repos/microsoft/azure-pipelines-tasks/commits/main")
        {
            Ok(revision.clone().into_bytes())
        } else if request.url.ends_with(&format!("/tar.gz/{revision}")) {
            Ok(b"azure task archive".to_vec())
        } else if request
            .url
            .ends_with(&format!("/{revision}/Tasks/UsePythonVersionV0/task.json"))
        {
            Ok(b"{\"execution\":{\"Node20_1\":{\"target\":\"main.js\"}}}".to_vec())
        } else {
            Err(format!("unexpected URL {}", request.url))
        }
    };

    let fetched = resolve_dependency(
        &dependency(Provider::Azure, DependencyKind::Task, "UsePythonVersion@0"),
        &mut get,
    )
    .expect("resolve Azure task");

    assert_eq!(fetched.revision, revision);
    assert!(fetched.source.ends_with("/Tasks/UsePythonVersionV0"));
    assert_eq!(
        fetched
            .semantic_source
            .as_ref()
            .map(|value| value.0.as_str()),
        Some("Tasks/UsePythonVersionV0/task.json")
    );
    assert_eq!(requests.len(), 3);
}

#[test]
fn circleci_orb_requires_an_exact_production_version_and_fetches_source() {
    let version_id = "7a09fb7b-4415-4aee-bc0f-a2f7f8395824";
    let mut requests = Vec::new();
    let mut get = |request: &ResolverRequest| {
        requests.push(request.url.clone());
        if request.url.contains("/api/v3/orb/versions?") {
            Ok(format!(
                "{{\"data\":[{{\"attributes\":{{\"version\":\"5.0.3\"}},\"id\":\"{version_id}\"}}]}}"
            )
            .into_bytes())
        } else if request.url.ends_with(&format!("/{version_id}/source")) {
            Ok(b"version: 2.1\ncommands: {}\n".to_vec())
        } else {
            Err(format!("unexpected URL {}", request.url))
        }
    };

    let fetched = resolve_dependency(
        &dependency(
            Provider::Circleci,
            DependencyKind::Orb,
            "circleci/node@5.0.3",
        ),
        &mut get,
    )
    .expect("resolve CircleCI orb");

    assert_eq!(fetched.revision, "5.0.3");
    assert_eq!(
        fetched
            .semantic_source
            .as_ref()
            .map(|value| value.0.as_str()),
        Some(".circleci/config.yml")
    );
    assert!(
        resolve_dependency(
            &dependency(Provider::Circleci, DependencyKind::Orb, "circleci/node@5",),
            &mut get,
        )
        .is_err()
    );
    assert_eq!(requests.len(), 2);
}

#[test]
fn gitlab_repository_file_uses_the_declared_project_revision_and_path() {
    let revision = "d".repeat(40);
    let mut requests = Vec::new();
    let mut get = |request: &ResolverRequest| {
        requests.push(request.url.clone());
        if request
            .url
            .contains(&format!("/repository/commits/{revision}"))
        {
            Ok(format!("{{\"id\":\"{revision}\"}}").into_bytes())
        } else if request.url.contains("archive.tar.gz?sha=") {
            Ok(b"project include archive".to_vec())
        } else if request
            .url
            .contains("repository/files/templates%2Fbuild.yml/raw")
        {
            Ok(b"build:\n  script: echo exact\n".to_vec())
        } else {
            Err(format!("unexpected URL {}", request.url))
        }
    };
    let dependency = located_dependency(
        Provider::Gitlab,
        DependencyKind::Repository,
        &format!("acme/ci:/templates/build.yml@{revision}"),
        DependencyLocator::RepositoryFile {
            repository: "acme/ci".to_owned(),
            revision: Some(revision.clone()),
            path: "/templates/build.yml".to_owned(),
            repository_type: None,
        },
    );

    let fetched = resolve_dependency(&dependency, &mut get).expect("resolve GitLab include");

    assert_eq!(fetched.revision, revision);
    assert!(
        fetched
            .source
            .ends_with(&format!("/{}/templates/build.yml", fetched.revision))
    );
    assert_eq!(
        fetched
            .semantic_source
            .as_ref()
            .map(|source| source.0.as_str()),
        Some("templates/build.yml")
    );
    assert_eq!(requests.len(), 3);
}

#[test]
fn azure_repository_locators_support_github_and_native_repositories() {
    let github_revision = "e".repeat(40);
    let mut github_requests = Vec::new();
    let mut github_get = |request: &ResolverRequest| {
        github_requests.push(request.url.clone());
        if request
            .url
            .contains("/repos/org/templates/commits/refs%2Ftags%2Fv1")
        {
            Ok(github_revision.clone().into_bytes())
        } else if request.url.ends_with(&format!("/tar.gz/{github_revision}")) {
            Ok(b"GitHub template archive".to_vec())
        } else if request
            .url
            .ends_with(&format!("/{github_revision}/shared.yml"))
        {
            Ok(b"steps:\n  - script: echo exact\n".to_vec())
        } else {
            Err(format!("unexpected URL {}", request.url))
        }
    };
    let github = located_dependency(
        Provider::Azure,
        DependencyKind::Template,
        "shared.yml@shared",
        DependencyLocator::RepositoryFile {
            repository: "org/templates".to_owned(),
            revision: Some("refs/tags/v1".to_owned()),
            path: "shared.yml".to_owned(),
            repository_type: Some("github".to_owned()),
        },
    );
    let fetched = resolve_dependency(&github, &mut github_get).expect("resolve GitHub alias");
    assert_eq!(fetched.revision, github_revision);
    assert_eq!(github_requests.len(), 3);

    let native_revision = "f".repeat(40);
    let mut native_requests = Vec::new();
    let mut native_get = |request: &ResolverRequest| {
        native_requests.push(request.url.clone());
        if request.url.contains("/commits?") {
            Ok(format!("{{\"value\":[{{\"commitId\":\"{native_revision}\"}}]}}").into_bytes())
        } else if request.url.contains("recursionLevel=Full") {
            Ok(b"Azure Repos archive".to_vec())
        } else if request.url.contains("path=%2Fshared.yml") {
            Ok(b"steps:\n  - script: echo exact\n".to_vec())
        } else {
            Err(format!("unexpected URL {}", request.url))
        }
    };
    let native = located_dependency(
        Provider::Azure,
        DependencyKind::Template,
        "shared.yml@templates",
        DependencyLocator::RepositoryFile {
            repository: "org/project/templates".to_owned(),
            revision: Some("refs/tags/v1".to_owned()),
            path: "shared.yml".to_owned(),
            repository_type: Some("azurereposgit".to_owned()),
        },
    );
    let fetched = resolve_dependency(&native, &mut native_get).expect("resolve Azure Repos alias");
    assert_eq!(fetched.revision, native_revision);
    assert_eq!(
        fetched
            .semantic_source
            .as_ref()
            .map(|source| source.0.as_str()),
        Some("shared.yml")
    );
    assert_eq!(native_requests.len(), 3);
}

#[test]
fn container_images_resolve_tags_and_redact_registry_tokens_from_debug_output() {
    let token = "registry-secret-value";
    let mut requests = Vec::new();
    let mut get = |request: &ResolverRequest| {
        assert!(!format!("{request:?}").contains(token));
        requests.push(request.clone());
        if request.url.starts_with("https://auth.docker.io/token?") {
            Ok(format!("{{\"token\":\"{token}\"}}").into_bytes())
        } else if request
            .url
            .starts_with("https://registry-1.docker.io/v2/library/alpine/manifests/")
        {
            assert!(request.headers.iter().any(
                |(name, value)| name == "Authorization" && value == &format!("Bearer {token}")
            ));
            Ok(b"immutable manifest".to_vec())
        } else {
            Err(format!("unexpected URL {}", request.url))
        }
    };
    let tagged = dependency(
        Provider::Circleci,
        DependencyKind::ContainerImage,
        "alpine:3.20",
    );

    let fetched = resolve_dependency(&tagged, &mut get).expect("resolve image tag");

    assert!(fetched.revision.starts_with("sha256:"));
    assert_eq!(fetched.content, b"immutable manifest");
    assert!(
        fetched
            .source
            .starts_with("oci://registry-1.docker.io/library/alpine@sha256:")
    );
    let digest = format!("sha256:{}", "1".repeat(64));
    let pinned = dependency(
        Provider::Github,
        DependencyKind::ContainerImage,
        &format!("docker://alpine@{digest}"),
    );
    let fetched = resolve_dependency(&pinned, &mut get).expect("accept pinned image");
    assert_eq!(fetched.revision, digest);
    assert_eq!(requests.len(), 2, "pinned image must stay offline");
}
